library/src/main/java/com/novoda/storagepathfinder/AndroidDeviceStorageInspector.java

```diff
@@ -28,63 +29,78 @@ public AndroidDeviceStorageInspector(
             AndroidSystem androidSystem
     ) {
         this.fileSystem = fileSystem;
-        String primaryStoragePath = getPrimaryStoragePath();
-        String basePath = getDirectoryPathAboveTheAndroidFolderFrom(primaryStoragePath);    // TODO: try to remove the duplicated code.
-        secondaryStorageInspectors.add(new ExternalFileDirectoryInspector(context, deviceFeatures, basePath));
+        CommonDirectories commonDirectories = fileSystem.getCommonDirectories();
+        primaryStorageInspector = new ExternalDirectoryPrimaryStorageInspector(commonDirectories);
 
+        StoragePath primaryBasePath = primaryStorageInspector.getPrimaryDeviceStorageBasePath();
         secondaryStorageInspectors.add(new EnvironmentVariableStorageInspector(androidSystem));
+        secondaryStorageInspectors.add(new ExternalFileDirectoryInspector(context, deviceFeatures, primaryBasePath.asString()));
     }
 
     @Override
-    public List<DeviceStorageRoot> getDeviceStorageRoots() {
-        List<DeviceStorageRoot> storageRoots = new ArrayList<>(2);
+    public StoragePath getPrimaryStorageBasePath() {
+        return primaryStorageInspector.getPrimaryDeviceStorageBasePath();
 
-        String primaryStoragePath = getPrimaryStoragePath();
-        String basePath = getDirectoryPathAboveTheAndroidFolderFrom(primaryStoragePath);
-        DeviceStorageRoot primaryStorageRoot = new DeviceStorageRoot(basePath, getPrimaryStoragePath(), DeviceStorageRoot.Type.PRIMARY);
-        storageRoots.add(primaryStorageRoot);
+    }
+
+    @Override
+    public StoragePath getPrimaryStorageApplicationPath() {
+        return primaryStorageInspector.getPrimaryDeviceStorageApplicationPath();
+    }
 
-        Set<DeviceStorageRoot> secondaryStorageRoots = findActiveSecondaryStorageRoots();
+    @Override
+    public List<StoragePath> getSecondaryStorageBasePath() {
+        List<StoragePath> storageRoots = new ArrayList<>(2);
```

Comment:
This should use `secondaryStorageInspectors.size` no? Otherwise if you add another to the list this will throw.

Reply:
Nope. 
If you pass a capacity to an ArrayList it sets the initial capacity, and if you add more elements it will then do some extra internal restructuring to increase the size. 
I will remove the 2 anyway though, as 95% of the time this will be 1.

Reply:
Very true, doing a restructure is still bad 😄